### ISSUE TITLE
DeepEP v2 doesn't work with per-layer compile

### ISSUE THREAD

**Issue by wwwjn** [pytorch/torchtitan#3830]
### Bug description

   Config: qwen3_moe_deepep (debugmodel_moe, moe_comm_backend=deepep, EP=4), --compile.enable 
  --debug.seed 42 --training.steps 5. It hard-errors at trace time on step 1:

  torch._dynamo.exc.Unsupported: torch.* op returned non-Tensor
    Higher Order Operator: torch.utils.checkpoint.checkpoint
    Developer debug context: example_value type: int; op: call_function;
      target: <built-in function _len_torch_dispatch_stack>     (gb0208)
  
##  Root cause (torchtitan/distributed/deepep/deepep.py:515):

  with _disable_current_modes():          # <- dispatch_tokens()
      get_buffer(group, hidden=..., num_max_tokens_per_rank=..., num_topk=...)
  
  _disable_current_modes() calls _len_torch_dispatch_stack(), which returns a Python int (a
  non-Tensor). Dynamo can't put a non-Tensor torch-op result into the FX graph, so it tries to
  graph-break there.

  Why it's a hard error and not a soft break: qwen3_moe_deepep uses SelectiveAC, so each compiled
  block sits inside the torch.utils.checkpoint.checkpoint higher-order operator — and graph breaks
  are illegal inside a HOP, so the break is promoted to an exception.
  
  The irony: that exact _disable_current_modes() is there specifically to make SAC work
  (deepep.py:513-514 comment — it hides the all_gather_object/aten._to_copy MUST_SAVE op from SAC's
  __torch_dispatch__). So:
  - SAC + DeepEP, eager → works (this is the verified numerics-parity path)
  - SAC + DeepEP + per-layer compile → collides; the same mode-disable that makes SAC work breaks
  Dynamo tracing.

  Notes:
  - This is not the cu130 DTensor codegen bug I was watching for DeepEP-v1 ("DeviceMesh cannot be interpreted
  as an integer") — the trace never gets that far; it fails earlier at the dispatch mode-disable.
  - The eager (no --compile.enable) run of this exact config works fine — DeepEP v2 training
  numerics parity is already verified.

### Versions

latest main

**Comment by syed-ahmed** [comment]
@wwwjn I've seen this before on the hybrid ep path. I think _disable_current_modes shouldn't be used and the get_buffer call should just be moved inside the custom op impl, such that torch.compile doesn't see it: https://github.com/pytorch/torchtitan/pull/3360#issuecomment-4455313067

**Comment by tianyu-l** [comment]
@syed-ahmed 

> get_buffer call should just be moved inside the custom op impl

Could we follow similar pattern to this MinimalAsyncEP? https://github.com/pytorch/torchtitan/blob/main/torchtitan/distributed/minimal_async_ep/api.py#L176

I believe it's compile friendly.
